### PR TITLE
fix(language): resolve Python closure variables in DSL function bodies

### LINKED ISSUE #276: [Bug] DSL parser raises UndefinedVariableError for Python global variables used as function call arguments
### Component

Parser

### Description

The DSL parser raises `UndefinedVariableError` when a Python global variable is used as a function call argument inside a `@pl.function` DSL function body. The parser's `parse_name` method resolves `ast.Name` nodes exclusively through `scope_manager.lookup_var()`, with no fallback to Python globals or the enclosing scope.

Notably, the same global variables **work correctly in type annotations** (e.g., `pl.Tensor[TENSOR_SHAPE, DTYPE]`), because annotations are evaluated by Python directly and never go through the DSL parser.

Related: #200

### Steps to Reproduce

1. Define a Python global variable:
   ```python
   OFFSET = [0, 0]
   TILE_SHAPE = [32, 32]
   ```
2. Reference it as a function call argument inside a `@pl.program` / `@pl.function` body:
   ```python
   @pl.program
   class Program:
       @pl.function(type=pl.FunctionType.InCore)
       def main(self, a: pl.Tensor[[128, 128], pl.FP32], output: pl.Tensor[[128, 128], pl.FP32]):
           tile_a: pl.Tile[[32, 32], pl.FP32] = pl.load(a, OFFSET, TILE_SHAPE)
           ...
   str(Program)
   ```
3. Observe the error.

### Expected Behavior

The parser should resolve Python global variables from the enclosing Python scope when they appear as function call arguments inside the DSL function body, or produce a clear diagnostic message explaining that global variables are not supported in this position and suggesting the use of inline literals instead.

### Actual Behavior

```
pypto.language.parser.diagnostics.exceptions.UndefinedVariableError: Undefined variable 'OFFSET'
```

The parser fails immediately upon encountering the first `ast.Name` node (`OFFSET`) that is not defined in the DSL scope.

### Git Commit ID

382d787ede3e4eecfb6352f6b01358384891026f

### NPU Kind

N/A (not hardware-specific)

### Host Platform

macOS (aarch64)

### Additional Context

- `ast_parser.py` line ~1064: `scope_manager.lookup_var(var_name)` returns `None` for any name not introduced by the DSL itself; no Python globals lookup is attempted.
- Workaround: replace global variable references in function call arguments with inline literals (e.g., `[0, 0]` instead of `OFFSET`). Global variables still work in type annotations.

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

Fixes #276. `parse_name()` previously raised `UndefinedVariableError` when Python closure/global variables were used as **positional arguments** in `@pl.function` bodies (e.g., `pl.block.load(t, OFFSET, TILE_SHAPE)`). The same variables already worked in type annotations and kwargs via `ExprEvaluator`.

- Add `python_value_to_ir()` and `try_eval_as_ir()` to `ExprEvaluator` for converting Python values (`int`, `float`, `bool`, `list`, `tuple`, `ir.Expr`) to IR expressions
- Modify `parse_name()` to fall back to closure resolution when a name isn't in the DSL scope
- Add 8 tests covering all supported types, shadowing, and error cases
- Document closure variable resolution in Python syntax spec

## Testing

- [x] 8 new tests in `test_closure_var_resolution.py` pass
- [x] All 327 parser tests pass (no regressions)
- [x] Full unit test suite passes (1589/1589)
- [x] All pre-commit hooks pass (ruff, pyright, markdownlint, etc.)

### CODE REVIEW

File: python/pypto/language/parser/expr_evaluator.py [hunk 1]
```diff
@@ -115,6 +114,53 @@ def try_eval_expr(self, node: ast.expr) -> tuple[bool, Any]:
...
+        """
+        # bool before int because isinstance(True, int) is True
+        if isinstance(value, bool):
+            return ir.ConstBool(value, span)
+        if isinstance(value, int):
+            return ir.ConstInt(value, DataType.DEFAULT_CONST_INT, span)
+        if isinstance(value, float):
+            return ir.ConstFloat(value, DataType.DEFAULT_CONST_FLOAT, span)
+        if isinstance(value, ir.Expr):
+            return value
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/expr_evaluator.py:136]: Fixed in 8261178 — added DynVar support to `python_value_to_ir`, converting to `ir.Var` with `INDEX` scalar type, consistent with `TypeResolver`.

File: tests/ut/language/parser/test_closure_var_resolution.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+        @pl.function
+        def func(
+            t: pl.Tensor[[128, 128], pl.FP32], out: pl.Tensor[[128, 128], pl.FP32]
+        ) -> pl.Tensor[[128, 128], pl.FP32]:
+            a: pl.Tile[[64, 64], pl.FP32] = pl.block.load(t, OFFSET, TILE_SHAPE)
+            result: pl.Tensor[[128, 128], pl.FP32] = pl.block.store(a, OFFSET, TILE_SHAPE, output_tensor=out)
+            return result
+
+        assert isinstance(func, ir.Function)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/language/parser/test_closure_var_resolution.py:86]: Fixed in 8261178 — added `test_nested_list_closure_var` that verifies recursive conversion of `[[0, 0], [64, 64]]` into nested MakeTuple via `pl.block.load`.

File: python/pypto/language/parser/expr_evaluator.py [hunk 3]
```diff
@@ -115,6 +114,53 @@ def try_eval_expr(self, node: ast.expr) -> tuple[bool, Any]:
...
+        Returns:
+            IR expression, or None if the name cannot be resolved.
+
+        Raises:
+            ParserTypeError: If the value is resolved but has an unsupported type.
+        """
+        success, value = self.try_eval_expr(node)
+        if not success:
+            return None
+        return self.python_value_to_ir(value, self._get_span(node))
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/expr_evaluator.py:165]: The feature is well-covered by 10 integration tests in `test_closure_var_resolution.py` that exercise all type paths end-to-end. Adding isolated unit tests for these thin methods would mostly duplicate that coverage. Skipping per team agreement.

File: tests/ut/language/parser/test_closure_var_resolution.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/language/parser/test_closure_var_resolution.py:86]: Empty lists/tuples aren't a realistic use case for positional args in DSL function calls (e.g., `pl.block.load(t, [], [])`), and `MakeTuple([])` construction is already covered by the IR layer. Skipping per team agreement.
